### PR TITLE
Specification updates.

### PR DESCRIPTION
**Pull request by grp**

### CODE REVIEW

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp [hunk 1]
```diff
@@ -195,6 +219,61 @@ parse(Context *context, plist::Dictionary const *dict, std::unordered_set<std::s
...
+        }
+    }
+
+    if (AFTs != nullptr) {
+        _allowedFileTypes = std::unordered_map<std::string, std::vector<std::string>>();
+        for (size_t n = 0; n < AFTs->count(); n++) {
+            if (auto AFT = AFTs->value<plist::Array>(n)) {
+                auto result = _allowedFileTypes->insert({ AFTs->key(n), std::vector<std::string>() });
+
+                for (size_t m = 0; m < AFT->count(); m++) {
```
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Looks verbose; can you just construct the vector first, then insert everything after the secondary loop?

At the very least, move the vector* fetching outside of the secondary loop.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp [hunk 2]
```diff
@@ -195,6 +219,61 @@ parse(Context *context, plist::Dictionary const *dict, std::unordered_set<std::s
...
+                    }
+                }
+            }
+        }
+    }
+
+    if (BPFRAs != nullptr) {
+        _buildPhaseFileRefAdditions = std::unordered_map<std::string, std::vector<FileReference>>();
+        for (size_t n = 0; n < BPFRAs->count(); n++) {
+            if (auto BPFRA = BPFRAs->value<plist::Array>(n)) {
```
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: For every failure to cast, should you print an error saying the format is bad?

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp [hunk 3]
```diff
@@ -195,6 +219,61 @@ parse(Context *context, plist::Dictionary const *dict, std::unordered_set<std::s
...
+                for (size_t m = 0; m < BPFRA->count(); m++) {
+                    if (auto FRA = BPFRA->value<plist::Dictionary>(m)) {
+                        auto result = _buildPhaseFileRefAdditions->insert({ BPFRAs->key(n), std::vector<FileReference>() });
+
+                        FileReference fileReference;
+                        if (fileReference.parse(FRA)) {
```
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Same as other comment:  it'd be nice to construct the vector std::vector<FileReference()> first, or move the fileRefAdditions initialization out of the inner loop.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp [hunk 4]
```diff
@@ -55,28 +55,36 @@ parse(Context *context, plist::Dictionary const *dict, std::unordered_set<std::s
...
+    auto AWCR   = unpack.coerce <plist::Boolean> ("AddWatchCompanionRequirement");
+    auto ROP    = unpack.coerce <plist::Boolean> ("RunsOnProxy");
+    auto DSA    = unpack.coerce <plist::Boolean> ("DisableSchemeAutocreation");
+    auto VEB    = unpack.coerce <plist::Boolean> ("ValidateEmbeddedBinaries");
+    auto SODR   = unpack.coerce <plist::Boolean> ("SupportsOnDemandResources");
+    auto CEASL  = unpack.coerce <plist::Boolean> ("CanEmbedAddressSanitizerLibraries");
+    auto CECSL  = unpack.coerce <plist::Boolean> ("CanEmbedCompilerSanitizerLibraries");
+    auto PPR    = unpack.coerce <plist::Boolean> ("ProvisioningProfileRequired");
+    auto PPS    = unpack.coerce <plist::Boolean> ("ProvisioningProfileSupported");
     auto RSEF  = unpack.coerce <plist::String> ("RunpathSearchPathForEmbeddedFrameworks");
```
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Nit: fix the alignment at this one line

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp [hunk 5]
```diff
@@ -283,3 +370,31 @@ Check(ext::optional<std::string> const &check, ext::optional<std::string> const
...
+parse(plist::Dictionary const *dict)
+{
+    std::unordered_set<std::string> seen;
+    auto unpack = plist::Keys::Unpack("FileReference", dict, &seen);
+
+    auto RVN = unpack.cast <plist::String> ("RegionVariantName");
+    auto P   = unpack.cast <plist::String> ("Path");
+
+    if (!unpack.complete(true)) {
+        fprintf(stderr, "%s", unpack.errorText().c_str());
```
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: return false?

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 2 above)
Comment (grp) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Yeah, in theory. In practice, don't really have error reporting setup in a way that won't get annoying.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 1 above)
Comment (grp) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Much better, thanks.

File: Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h [hunk 6]
```diff
@@ -65,8 +65,29 @@ class ProductType : public Specification {
         bool parse(plist::Dictionary const *dict);
     };
 
+public:
```
Comment (grp) [inline Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h]: A specification is a `.xcspec` file that defines various things, like available build settings, compiler flags, or known file types. A product type is a specification that defines a kind of thing built by a target, like an "Application" or a "Static Library". In the Xcode UI they're under the "+" icon in the list of targets:

screen shot 2016-10-13 at 11 22 03

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 5 above)
Comment (grp) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Not an error, just warning about unknown keys.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 2 above)
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Open up an issue and add some TODOs then please.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 5 above)
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Ok.  Prefix with "Warning: " then?

File: Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h (hunk 6 above)
Comment (Ktwu) [inline Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h]: Cool thanks.

Followup question: is there an external resource I can read that will help educate me on how xcodebuild builds things so I don't have to ask you as many stupid questions? :)

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 5 above)
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Yah, you're right.

File: Libraries/pbxspec/Sources/PBX/ProductType.cpp (hunk 2 above)
Comment (Ktwu) [inline Libraries/pbxspec/Sources/PBX/ProductType.cpp]: Got it.

File: Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h [hunk 7]
```diff
@@ -65,8 +65,29 @@ class ProductType : public Specification {
...
+        FileReference();
+
+    public:
+        inline ext::optional<std::string> const &regionVariantName() const
+        { return _regionVariantName; }
+        inline ext::optional<pbxsetting::Value> const &path() const
+        { return _path; }
+
+    protected:
+        bool parse(plist::Dictionary const *dict);;
```
Comment (sas) [inline Libraries/pbxspec/Headers/pbxspec/PBX/ProductType.h]: What does this mean? What change are you suggesting?

**Pull request merged**